stream.go

```diff
@@ -704,6 +698,18 @@ func (cs *clientStream) withRetry(op func(a *csAttempt) error, onSuccess func())
 			// already be status errors.
 			return toRPCErr(op(cs.attempt))
 		}
+		if len(cs.buffer) == 0 {
+			// For the first op, which controls creation of the stream and
+			// assigns cs.attempt, we need to create a new attempt inline
+			// before executing the first op.  On subsequent ops, the attempt
+			// is created immediately before replaying the ops.
+			var err error
+			if cs.attempt, err = cs.newAttemptLocked(false /* isTransparent */); err != nil {
+				cs.mu.Unlock()
+				cs.finish(err)
```

Comment:
`cs.finish()` grabs the lock again. Can something happen between us calling `Unlock()` on line 708 and `cs.finish()` calling `Lock()` that can cause trouble?

Reply:
Since this is only for the first op, the `clientStream` we're using is only visible to the caller at this point (`newClientStream`).